File: dna_fsdll/DnaSample.fs
module DnaSample

open System.Diagnostics
open ExcelDna.Integration

let range = [ for r in [1..100] do
              for c in [1..100] -> (r, c) ]

open Util.NonCached

[<ExcelCommand(MenuName="Bench(F#)", MenuText="COM Object without Cache")>]
let nonCachedCom () =
  let sw = Stopwatch.StartNew()

  let excel = ExcelDnaUtil.Application
  let sheet = excel?ActiveSheet
  range
  |> List.iter (fun (row, col) ->
       let cell = sheet?Cells?Item(row, col)
       cell?Value2 <- 1
     ) 

  sw.Stop()
  excel?StatusBar <- sw.Elapsed.TotalSeconds

open Util.Cached

[<ExcelCommand(MenuName="Bench(F#)", MenuText="COM Object with Cache")>]
let cachedCom () =
  let sw = Stopwatch.StartNew()

  let excel = ExcelDnaUtil.Application
  let sheet = excel?ActiveSheet
  range
  |> List.iter (fun (row, col) ->
       let cell = sheet?Cells?Item(row, col)
       cell?Value2 <- 1
     ) 

  sw.Stop()
  excel?StatusBar <- sw.Elapsed.TotalSeconds

[<ExcelCommand(MenuName="Bench(F#)", MenuText="DNA Object")>]
let dna () =
  let sw = Stopwatch.StartNew()

  range
  |> List.iter (fun (row, col) ->
       let row, col = row - 1, col - 1
       let cell = ExcelReference(row, col)
       cell.SetValue(1) |> ignore
     ) 

  sw.Stop()
  ExcelDnaUtil.Application?StatusBar <- sw.Elapsed.TotalSeconds

File: dna_csdll/DnaSample.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using ExcelDna.Integration;

namespace dna_csdll
{
    public static class DnaSample
    {
        static IEnumerable<Tuple<int, int>> range =
            from r in Enumerable.Range(1, 99)
            from c in Enumerable.Range(1, 99)
            select Tuple.Create(r, c);

        [ExcelCommand(MenuName = "Bench(C#)", MenuText = "COM Object")]
        public static void ComCs()
        {
            var sw = Stopwatch.StartNew();

            dynamic excel = ExcelDnaUtil.Application;
            dynamic sheet = excel.ActiveSheet;
            foreach (var r_c in range)
            {
                dynamic cell = sheet.Cells.Item(r_c.Item1, r_c.Item2);
                cell.Value2 = 1;
            }

            sw.Stop();
            excel.StatusBar = sw.Elapsed.TotalSeconds;
        }

        [ExcelCommand(MenuName = "Bench(C#)", MenuText = "DNA Object")]
        public static void DnaCS()
        {
            var sw = Stopwatch.StartNew();

            foreach (var r_c in range)
            {
                var row = r_c.Item1 - 1;
                var col = r_c.Item2 - 1;
                var cell = new ExcelReference(row, col);
                cell.SetValue(1);
            }

            sw.Stop();
            dynamic excel = ExcelDnaUtil.Application;
            excel.StatusBar = sw.Elapsed.TotalSeconds;
        }
    }
}